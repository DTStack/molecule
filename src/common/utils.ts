export function searchById(id) {
    return (item) => item.id === id;
}

/**
 * Clone a new object by an object instance
 * @param origin Original object instance
 */
export function cloneInstance<T>(origin: T) {
    try {
        const prototypes = Object.getPrototypeOf(origin) || {};
        Object.keys(prototypes).forEach((prop) => {
            if (typeof prototypes[prop] === 'function') {
                prototypes[prop].bind(origin);
            }
        });
        return {
            ...origin,
            ...prototypes,
        };
    } catch (e) {
        console.error('Function cloneInstance error:', e);
    }
}

/**
 * Merge multiple functions to one function
 * @param funcs
 */
export function mergeFunctions(...funcs) {
    return function (...args) {
        funcs.filter((fn) => !!fn).forEach((fn) => fn?.(...args));
    };
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function mergeObjects<T>(source: object, target: object): T {
    for (const prop in target) {
        if (typeof source[prop] === 'object' && target[prop]) {
            source[prop] = mergeObjects(source[prop], target[prop]);
        } else {
            source[prop] = target[prop];
        }
    }
    return source as any;
}

/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: object): object {
    const flatted = {};
    const recurse = (parent: string, target: object) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: object): object {
    let normalized = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = mergeObjects(normalized, root);
        }
    }
    return normalized;
}

/**
 * Determine if a color is light or dark.
 * @param color HEX or RGB
 */
export function colorLightOrDark(color: string) {
    // Variables for red, green, blue values
    let r: number;
    let g: number;
    let b: number;

    // Check the format of the color, HEX or RGB?
    if (color.match(/^rgb/)) {
        // If RGB --> store the red, green, blue values in separate variables
        const matchArray =
            color.match(
                /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
            ) || [];
        r = +matchArray[1];
        g = +matchArray[2];
        b = +matchArray[3];
    } else {
        // If hex --> Convert it to RGB
        let rgbNum = +('0x' + color.slice(1, 7));
        if (color.length < 5) {
            rgbNum = +('0x' + color.slice(1).replace(/./g, '$&$&').slice(0, 6));
        }
        r = rgbNum >> 16;
        g = (rgbNum >> 8) & 255;
        b = rgbNum & 255;
    }

    // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
    const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b));

    // Using the HSP value, determine whether the color is light or dark
    if (hsp > 127.5) {
        return 'light';
    }
    return 'dark';
}
