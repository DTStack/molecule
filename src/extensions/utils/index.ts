import type { IMenuItemProps, IterableItem } from 'mo/types';

export function createMenuDuplicate(item: IMenuItemProps): IMenuItemProps {
    return {
        ...item,
        id: `${item.id}_duplicate`,
        name: `隐藏 “${item.name}”`,
        icon: undefined,
        symbolic: item.id,
    };
}

export function createContextMenu(items: IterableItem[]): IMenuItemProps[] {
    return items.map((item) => ({
        id: item.id,
        name: item.name,
        disabled: item.disabled,
        icon: item.hidden ? undefined : 'check',
    }));
}

export function isInputEle() {
    return document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
}

/**
 * Judge current active element is monaco-editor's textarea or native element
 */
export function isNativeWidget() {
    const isEditableEle =
        document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
    if (!isEditableEle) return true;
    return !document.activeElement?.classList.contains('inputarea');
}
