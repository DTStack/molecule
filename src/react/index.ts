export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
