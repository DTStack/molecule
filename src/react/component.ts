import { cloneDeep } from 'lodash';
import { EventEmitter, GlobalEvent } from 'mo/common/event';

export enum ComponentEvents {
    Update = 'Component.Update',
}

export interface IComponent<S = any> {
    /**
     * Set the Component state
     * @param values The next values of state
     * @param callback calling after setState
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Trigger the Component update event
     * @param nextState
     */
    render(nextState?: S): void;
    /**
     * Listen to the Component state update event
     * @param listener
     */
    onUpdateState(listener: (prevState: S, nextState: S) => void): void;
    /**
     * Remove the Component update event listening, default is remove all,
     * also you can remove one by pass the listener
     */
    removeOnUpdateState(listener?: Function): void;
    /**
     * Force to update the Component
     */
    forceUpdate(): void;
    /**
     * Get the Component state
     */
    getState(): S;
}

export abstract class Component<S = any>
    extends GlobalEvent
    implements IComponent<S>
{
    protected abstract state: S;
    private _event: EventEmitter;

    constructor() {
        super();
        this._event = new EventEmitter();
    }

    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    public setState(
        values: Partial<S>,
        callback?: (prevState: S, nextState: S) => void
    ) {
        const nextState = Object.assign(this.state, values);
        this.render(nextState);
        callback?.(this.state, nextState);
    }

    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    public render(nextState?: S) {
        this._event.emit(ComponentEvents.Update, this.state, nextState);
    }

    public onUpdateState(listener: (prevState: S, nextState: S) => void) {
        this._event.subscribe(ComponentEvents.Update, listener);
    }

    public removeOnUpdateState(listener?: Function): void {
        this._event.unsubscribe(ComponentEvents.Update, listener);
    }

    public forceUpdate() {
        this.setState(cloneDeep(this.state));
    }

    public getState(): S {
        return this.state;
    }
}
