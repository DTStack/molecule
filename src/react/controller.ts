import { GlobalEvent } from 'mo/common/event';

export abstract class Controller extends GlobalEvent {
    public abstract initView(): void;
}
