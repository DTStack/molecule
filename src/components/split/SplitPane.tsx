import React, {
    ReactNode,
    useEffect,
    useMemo,
    useCallback,
    useRef,
    useState,
} from 'react';
import { cloneReactChildren } from 'mo/react';
import { classNames } from 'mo/common/className';
import { HTMLElementProps } from 'mo/common/types';
import Pane, { IPaneConfigs } from './pane';
import Sash from './sash';
import {
    paneItemClassName,
    splitClassName,
    splitDraggingClassName,
    splitVerticalClassName,
    splitHorizontalClassName,
    sashDisabledClassName,
    sashHorizontalClassName,
    sashVerticalClassName,
} from './base';

interface IAxis {
    x: number;
    y: number;
}

export interface ISplitProps extends HTMLElementProps {
    children: JSX.Element[];
    /**
     * Should allowed to resized
     *
     * default is true
     */
    allowResize?: boolean | boolean[];
    /**
     * Should show the sashes
     *
     * default is true
     */
    showSashes?: boolean | boolean[];
    /**
     * How to split the space
     *
     * default is vertical
     */
    split?: 'vertical' | 'horizontal';
    /**
     * Only support controlled mode, so it's required
     */
    sizes: (string | number)[];
    onChange: (sizes: number[]) => void;
    className?: string;
    sashClassName?: string;
    paneClassName?: string;
    /**
     * Specify the size fo resizer
     *
     * defualt size is 4px
     */
    resizerSize?: number;
}

/**
 * Convert size to absolute number or Infinity
 */
const assertsSize = function (
    size: string | number | undefined,
    sum: number,
    defaultValue = Infinity
) {
    if (typeof size === 'undefined') return defaultValue;
    if (typeof size === 'number') return size;
    if (size.endsWith('%')) return sum * (+size.replace('%', '') / 100);
    if (size.endsWith('px')) return +size.replace('px', '');
    return defaultValue;
};

const SplitPane = ({
    children,
    sizes: propSizes,
    allowResize: propAllowResize = true,
    showSashes = true,
    split = 'vertical',
    className,
    sashClassName,
    paneClassName,
    resizerSize = 4,
    onChange,
    ...restProps
}: ISplitProps) => {
    const axis = useRef<IAxis>({ x: 0, y: 0 });
    const wrapper = useRef<HTMLDivElement>(null);
    const [wrapperRect, setWrapperRect] = useState({});
    const [draging, setDrag] = useState(false);

    useEffect(() => {
        const resizeObserver = new ResizeObserver(() => {
            setWrapperRect(wrapper.current!.getBoundingClientRect());
        });
        resizeObserver.observe(wrapper.current!);
        return () => {
            resizeObserver.disconnect();
        };
    }, []);

    // Get some size infos via split
    const { sizeName, sPos, sAxis } = useMemo(
        function () {
            return {
                sizeName: split === 'vertical' ? 'width' : 'height',
                sPos: split === 'vertical' ? 'left' : 'top',
                sAxis: split === 'vertical' ? 'x' : 'y',
            };
        },
        [split]
    );

    const wrapSize: number = wrapperRect[sizeName] ?? 0;

    // Get limit sizes via children
    const paneLimitSizes = useMemo(
        function () {
            return children.map((childNode) => {
                const limits = [0, Infinity];
                if (childNode.type === Pane) {
                    const { minSize, maxSize } =
                        childNode.props as IPaneConfigs;
                    limits[0] = assertsSize(minSize, wrapSize, 0);
                    limits[1] = assertsSize(maxSize, wrapSize);
                }
                return limits;
            });
        },
        [children, wrapSize]
    );

    // perform the task for recalculating resizable
    const allowResize = useMemo(() => {
        if (typeof propAllowResize === 'boolean') {
            return new Array(children.length).fill(propAllowResize);
        }
        return children.map((_, index) => propAllowResize[index] ?? true);
    }, [children.length, propAllowResize]);

    /**
     * SplitPane allows sizes in string and number, but the state sizes only support number,
     * so convert string and number to number in here
     * ```ts
     * 'auto' -> divide the remaining space equally
     * 'xxxpx' -> xxx
     * 'xxx%' -> wrapper.size * xxx/100
     * xxx -> xxx
     * ```
     */
    const sizes = useMemo(
        function () {
            let count = 0;
            let curSum = 0;
            let allowResizeSum = 0;
            const res = children.map((_, index) => {
                const size = assertsSize(propSizes[index], wrapSize);
                if (size === Infinity) {
                    count++;
                } else {
                    curSum += size;
                    if (allowResize[index]) allowResizeSum += size;
                }
                return size;
            });

            const allowResizePanes: number = allowResize.filter(
                (item) => item === true
            ).length;

            if (allowResizePanes === 0) return res;

            // resize or illegal size input,recalculate pane sizes
            if (curSum > wrapSize || (!count && curSum < wrapSize)) {
                const gap = (curSum - wrapSize) / allowResizeSum;
                return res.map((size, index) => {
                    if (size === Infinity) return 0;
                    return allowResize[index] ? size - size * gap : size;
                });
            }

            if (count > 0) {
                const average = (wrapSize - curSum) / count;
                return res.map((size) => {
                    return size === Infinity ? average : size;
                });
            }

            return res;
        },
        [...propSizes, children.length, wrapSize, ...allowResize]
    );

    // Gets dragging axis position
    const sashPosSizes = useMemo(
        function () {
            return sizes.reduce(
                function (a, b) {
                    return [...a, a[a.length - 1] + b];
                },
                [0]
            );
        },
        [...sizes]
    );

    const onDragStart = useCallback(function (e) {
        axis.current = {
            x: e.pageX ?? e.screenX,
            y: e.pageY ?? e.screenY,
        };
        setDrag(true);
    }, []);

    const onDragging = useCallback(
        function (e, i) {
            const curAxis = {
                x: e.pageX ?? e.screenX,
                y: e.pageY ?? e.screenY,
            };
            let distanceX = curAxis[sAxis] - axis.current[sAxis];

            const leftBorder = -Math.min(
                sizes[i] - paneLimitSizes[i][0],
                paneLimitSizes[i + 1][1] - sizes[i + 1]
            );
            const rightBorder = Math.min(
                sizes[i + 1] - paneLimitSizes[i + 1][0],
                paneLimitSizes[i][1] - sizes[i]
            );

            if (distanceX < leftBorder) {
                distanceX = leftBorder;
            }
            if (distanceX > rightBorder) {
                distanceX = rightBorder;
            }

            const nextSizes = [...sizes];
            nextSizes[i] += distanceX;
            nextSizes[i + 1] -= distanceX;

            onChange?.(nextSizes);
        },
        [paneLimitSizes, onChange, sizes]
    );

    return (
        <div
            className={classNames(
                splitClassName,
                draging && splitDraggingClassName,
                split === 'vertical' && splitVerticalClassName,
                split === 'horizontal' && splitHorizontalClassName,
                className
            )}
            ref={wrapper}
            {...restProps}
        >
            {children.map((childNode, idx) => {
                const paneClasses = classNames(
                    paneItemClassName,
                    paneClassName
                );
                const paneStyle = {
                    [sizeName]: sizes[idx],
                    [sPos]: sashPosSizes[idx],
                };

                let sashChild: ReactNode = null;
                if (idx > 0) {
                    const disabled =
                        showSashes === false || showSashes?.[idx - 1] === false;
                    sashChild = (
                        <Sash
                            className={classNames(
                                disabled && sashDisabledClassName,
                                split === 'vertical'
                                    ? sashVerticalClassName
                                    : sashHorizontalClassName,
                                sashClassName
                            )}
                            style={{
                                [sizeName]: resizerSize,
                                [sPos]: sashPosSizes[idx] - resizerSize / 2,
                            }}
                            onDragStart={onDragStart}
                            onDragging={(e) => onDragging(e, idx - 1)}
                            onDragEnd={() => {
                                setDrag(false);
                            }}
                        />
                    );
                }

                if (childNode.type === Pane) {
                    const { className = '', style = {} } = childNode.props;
                    return (
                        <React.Fragment key={idx}>
                            {sashChild}
                            {cloneReactChildren(childNode, {
                                className: classNames(paneClasses, className),
                                style: { ...style, ...paneStyle },
                            })}
                        </React.Fragment>
                    );
                }

                return (
                    <React.Fragment key={idx}>
                        {sashChild}
                        <Pane className={paneClasses} style={paneStyle}>
                            {childNode}
                        </Pane>
                    </React.Fragment>
                );
            })}
        </div>
    );
};

export default SplitPane;
