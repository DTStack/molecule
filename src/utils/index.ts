import { Children, cloneElement, isValidElement } from 'react';
import { merge, omitBy } from 'lodash-es';
import type { editor } from 'mo/monaco';
import type { Arraylize, IColorTheme, IconType, IMenuItemProps, IterableItem, RecordWithId, UniqueId } from 'mo/types';

export function searchById<T extends RecordWithId<Record<string, any>>>(id?: UniqueId) {
    return (item: T) => item.id === id;
}

type sortIndexRequired = { sortIndex?: number; [key: string]: any };
/**
 * The smaller the sort number is, the more front the order is
 */
export function sortByIndex(a: sortIndexRequired, b: sortIndexRequired) {
    const prev = a.sortIndex || 0;
    const next = b.sortIndex || 0;
    return prev - next;
}

/**
 * Clone react children props
 * @param children React.ReactNode
 * @param props Parent props
 */
export function cloneReactChildren<P>(children: React.ReactNode, props: P): React.ReactNode {
    return Children.map(children, (child) => {
        if (isValidElement(child)) {
            return cloneElement(child, props);
        }
        return child;
    });
}

/**
 * Make an object to be obejct array
 */
export function arraylize<T>(data: Arraylize<T>) {
    return Array.isArray(data) ? data : [data];
}

/**
 * Classify arr into two parts
 */
export function classify<T>(arr: T[], predicate: (value: T, index: number) => boolean) {
    return classifyBy(arr, (...args) => Number(!predicate(...args)));
}

/**
 * Classify arr into multiply parts
 */
export function classifyBy<T>(arr: T[], predicate: (value: T, index: number) => number) {
    return arr.reduce<T[][]>((acc, cur, index) => {
        const idx = predicate(cur, index);
        acc[idx] ??= [];
        acc[idx].push(cur);
        return acc;
    }, []);
}

export function get<T extends RecordWithId<{ children?: T[] }>>(obj: Arraylize<T>, keyPath: string[]) {
    let root = { children: arraylize(obj) } as T;
    const stack = [...keyPath];
    while (stack.length) {
        const item = stack.shift();
        const next = root.children?.find((i) => i.id === item);
        if (!next) break;
        root = next;
    }
    return root as T;
}

/**
 * get the next icon
 * @param prev previous icon
 * @param checked
 */
export function toggleNextIcon(prev?: IconType, checked?: boolean) {
    if (typeof checked === 'boolean') {
        return checked === true ? 'check' : undefined;
    }
    if (typeof prev === 'string' || typeof prev === 'undefined') {
        return prev === 'check' ? undefined : 'check';
    }
    // JSX.Element doesn't change icon status
    return prev as IconType;
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function concatMenu(...args: IMenuItemProps[][]) {
    const nonEmptyArgs = args.filter((i) => Array.isArray(i) && i.length);
    if (!nonEmptyArgs.length) return [];
    return nonEmptyArgs.reduce((acc, cur) => {
        return acc.concat({ id: randomId(), type: 'divider' }, cur);
    });
}

/**
 * Converts an object of color values into CSS variables.
 *
 * @param {Record<string, string>} colors - The object containing color values.
 * @return {string} The CSS variables string.
 */
export function convertToCSSVars(colors: Record<string, string | null>): string {
    return `
        :root {
            ${Object.keys(colors)
                .map((id) => {
                    const color = colors[id];
                    if (!color) return '';
                    const colorName = id.replace('.', '-');
                    return `--${colorName}: ${color};`;
                })
                .filter(Boolean)
                .join('\n')}
        }
    `;
}

/**
 * Converts a token to a token theme rule.
 *
 * @param {IColorTheme['tokenColors']} token - The token to be converted.
 * @return {editor.ITokenThemeRule[]} The converted token theme rule.
 */
export function convertToToken(token?: IColorTheme['tokenColors']) {
    return token?.reduce<editor.ITokenThemeRule[]>((acc, cur) => {
        if (Array.isArray(cur.scope)) {
            cur.scope.forEach((s) => () => {
                if (!s) return;
                const target = acc.find((r) => r.token === s);
                if (target) {
                    Object.assign(target, {
                        ...cur.settings,
                    });
                } else {
                    acc.push({
                        token: s,
                        ...cur.settings,
                    });
                }
            });
        }
        return acc;
    }, []);
}

export function colorsToString(colors: Record<string, string | null>) {
    return omitBy(colors, (value) => !value) as Record<string, string>;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: Record<string, any>): Record<string, any> {
    let normalized: Record<string, any> = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root: Record<string, any> = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = merge(normalized, root);
        }
    }
    return normalized;
}
/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: Record<string, any>): Record<string, any> {
    const flatted: Record<string, any> = {};
    const recurse = (parent: string, target: Record<string, any>) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

export function getPrevOrNext<T>(arr: T[], idx: number) {
    return arr[idx - 1] ?? arr[idx + 1];
}

export function isElementInParentView(ele: HTMLElement, parent: HTMLElement) {
    const elRect = ele.getBoundingClientRect();
    const parentRect = parent.getBoundingClientRect();
    const left = elRect.left >= parentRect.left;
    const right = elRect.left + elRect.width <= parentRect.left + parentRect.width;
    const top = elRect.top >= parentRect.top;
    const bottom = elRect.top + elRect.height <= parentRect.top + parentRect.height;
    const overlay = left && right && top && bottom;
    let isWhichSide: 'right' | 'left' | 'top' | 'bottom' | undefined;
    if (!overlay) {
        if (!left) isWhichSide = 'left';
        else if (!right) isWhichSide = 'right';
        else if (!top) isWhichSide = 'top';
        else if (!bottom) isWhichSide = 'bottom';
    }
    return [overlay, isWhichSide] as const;
}

export function getNameForTitle(name: IterableItem['name']) {
    return typeof name === 'string' ? name : undefined;
}
