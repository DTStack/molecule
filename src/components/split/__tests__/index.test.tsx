import { sleep } from '@test/utils';
import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import {
    paneItemClassName,
    sashDisabledClassName,
    sashHoverClassName,
    sashItemClassName,
} from '../base';
import Pane from '../pane';
import SplitPane from '../SplitPane';

afterEach(cleanup);

describe('Test The SplitPane Component', () => {
    let original;
    beforeEach(() => {
        original = HTMLElement.prototype.getBoundingClientRect;
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = () => ({
            width: 500,
            height: 500,
        });
        global.ResizeObserver = jest.fn().mockImplementation((fn) => {
            fn();
            return {
                observe: jest.fn(),
                unobserve: jest.fn(),
                disconnect: jest.fn(),
            };
        });
    });

    afterEach(() => {
        HTMLElement.prototype.getBoundingClientRect = original;
    });

    test('Match Snapshot', () => {
        const { asFragment } = render(
            <SplitPane sizes={[100, 200]} onChange={jest.fn()}>
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('Match Snapshot in horizontal', () => {
        const { asFragment } = render(
            <SplitPane
                sizes={[100, 200]}
                split="horizontal"
                onChange={jest.fn()}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('Should NOT resizable', () => {
        const { container } = render(
            <SplitPane
                sizes={[100, 200]}
                showSashes={false}
                onChange={jest.fn()}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );

        const sashs = container.querySelectorAll(`.${sashItemClassName}`);

        expect(sashs.length).toBe(2);
        expect(
            Array.prototype.every.call(sashs, (sash: Element) =>
                sash.classList.contains(sashDisabledClassName)
            )
        ).toBeTruthy();
    });

    test('Should support each resizable', () => {
        const { container } = render(
            <SplitPane
                sizes={[20, 20]}
                showSashes={[false, true]}
                onChange={jest.fn()}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );

        const sashs = container.querySelectorAll(`.${sashItemClassName}`);

        expect(sashs.length).toBe(2);
        expect(sashs[0].classList.contains(sashDisabledClassName)).toBeTruthy();
        expect(sashs[1].classList.contains(sashDisabledClassName)).toBeFalsy();
    });

    test('Should convert "auto" and "xxxpx" and "xxx%" to absolute number ', () => {
        const { container } = render(
            <SplitPane
                sizes={['10%', '10px']}
                style={{ width: 500 }}
                onChange={jest.fn()}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );

        const panes = container.querySelectorAll<HTMLDivElement>(
            `.${paneItemClassName}`
        );

        expect(panes.length).toBe(3);
        expect(panes[0].style.width).toBe('50px');
        expect(panes[1].style.width).toBe('10px');
        expect(panes[2].style.width).toBe('440px');
    });

    test('Should support to adjust panes', () => {
        const mockFn = jest.fn();
        const { container, getByRole } = render(
            <SplitPane
                role="split"
                sizes={[100, 200]}
                style={{ width: 500 }}
                onChange={mockFn}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );

        const wrapper = getByRole('split');
        const sashs = container.querySelectorAll(`.${sashItemClassName}`);
        fireEvent.mouseDown(sashs[0], { screenX: 0 });
        fireEvent.mouseMove(wrapper, { screenX: 50 });
        fireEvent.mouseUp(wrapper);

        expect(mockFn).toBeCalled();
        expect(mockFn.mock.calls[0][0]).toEqual([150, 150, 200]);
    });

    test('Should with hover className in sashs', async () => {
        const { container } = render(
            <SplitPane
                role="split"
                sizes={['10%', '10px']}
                style={{ width: 500 }}
                onChange={jest.fn()}
            >
                <div>1</div>
                <div>2</div>
                <div>3</div>
            </SplitPane>
        );

        const sashs = container.querySelectorAll(`.${sashItemClassName}`);

        await act(async () => {
            fireEvent.mouseEnter(sashs[1]);
            await sleep(150);
        });

        expect(sashs[0].classList).not.toContain(sashHoverClassName);
        expect(sashs[1].classList).toContain(sashHoverClassName);

        fireEvent.mouseLeave(sashs[1]);
        expect(sashs[1].classList).not.toContain(sashHoverClassName);
    });

    test('Should support to resize', async () => {
        const mockFn = jest.fn();
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = () => ({
            width: 1000,
            height: 500,
        });
        const { container } = render(
            <SplitPane role="split" sizes={[250, 250]} onChange={mockFn}>
                <div>1</div>
                <div>2</div>
            </SplitPane>
        );

        const panes = container.querySelectorAll<HTMLDivElement>(
            `.${paneItemClassName}`
        );

        expect(panes.length).toBe(2);
        expect(panes[0].style.width).toBe('500px');
        expect(panes[1].style.width).toBe('500px');
    });

    test('Only the specified panel can be resized', async () => {
        const mockFn = jest.fn();
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = () => ({
            width: 1000,
            height: 500,
        });
        const { container } = render(
            <SplitPane
                role="split"
                sizes={[250, 250]}
                allowResize={[false]}
                onChange={mockFn}
            >
                <div>1</div>
                <div>2</div>
            </SplitPane>
        );

        const panes = container.querySelectorAll<HTMLDivElement>(
            `.${paneItemClassName}`
        );

        expect(panes.length).toBe(2);
        expect(panes[0].style.width).toBe('250px');
        expect(panes[1].style.width).toBe('750px');
    });

    test('Should have limited sizes', () => {
        const mockFn = jest.fn();
        const { container, getByRole } = render(
            <SplitPane role="split" sizes={[100, 200]} onChange={mockFn}>
                <Pane maxSize={150} minSize="50px">
                    <div>1</div>
                </Pane>
                <Pane maxSize={280}>
                    <div>2</div>
                </Pane>
                <div>3</div>
            </SplitPane>
        );

        const wrapper = getByRole('split');
        const sashs = container.querySelectorAll(`.${sashItemClassName}`);
        fireEvent.mouseDown(sashs[0], { screenX: 0 });
        fireEvent.mouseMove(wrapper, { screenX: -60 });

        expect(mockFn).toBeCalled();
        expect(mockFn.mock.calls[0][0]).toEqual([50, 250, 200]);

        fireEvent.mouseMove(wrapper, { screenX: 1000 });
        expect(mockFn.mock.calls[1][0]).toEqual([150, 150, 200]);

        fireEvent.mouseUp(wrapper);
    });
});
