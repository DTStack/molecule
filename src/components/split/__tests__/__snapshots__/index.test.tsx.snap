// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test The SplitPane Component Match Snapshot 1`] = `
<DocumentFragment>
  <div
    class="mo-split mo-split--vertical"
  >
    <div
      class="mo-split__pane"
      style="width: 100px; left: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="mo-split__sash mo-split__sash--vertical"
      role="Resizer"
      style="width: 4px; left: 98px;"
    />
    <div
      class="mo-split__pane"
      style="width: 200px; left: 100px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="mo-split__sash mo-split__sash--vertical"
      role="Resizer"
      style="width: 4px; left: 298px;"
    />
    <div
      class="mo-split__pane"
      style="width: 200px; left: 300px;"
    >
      <div>
        3
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test The SplitPane Component Match Snapshot in horizontal 1`] = `
<DocumentFragment>
  <div
    class="mo-split mo-split--horizontal"
  >
    <div
      class="mo-split__pane"
      style="height: 100px; top: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="mo-split__sash mo-split__sash--horizontal"
      role="Resizer"
      style="height: 4px; top: 98px;"
    />
    <div
      class="mo-split__pane"
      style="height: 200px; top: 100px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="mo-split__sash mo-split__sash--horizontal"
      role="Resizer"
      style="height: 4px; top: 298px;"
    />
    <div
      class="mo-split__pane"
      style="height: 200px; top: 300px;"
    >
      <div>
        3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
